OSF Preprints.js

```diff
@@ -0,0 +1,380 @@
+	for (let row of rows) {
+		let href = row.href;
+		let title = ZU.trimInternal(row.textContent);
+		if (!href || !title || !supportedSites.test(href)) continue;
+		if (checkOnly) return true;
+		found = true;
+		items[href] = title;
+	}
+	return found ? items : false;
+}
+
+function doWeb(doc, url) {
+	if (detectWeb(doc, url) == "multiple") {
+		Zotero.selectItems(getSearchResults(doc, false), function (items) {
+			if (items) ZU.doGet(constructAPIURL(Object.keys(items)), osfAPIImport);
+		});
+	}
+	else {
+		scrape(doc, url);
+	}
+}
+
+
+// takes and array of preprint URLs, extracts the ID and constructs an API call to OSF
+function constructAPIURL(urls) {
+	var ids = [];
+	for (let url of urls) {
+		let id;
+		if (url.match(/\.(io|com|org)\/([a-z0-9]+)/)) {
+			id = url.match(/\.(?:io|com|org)\/([a-z0-9]+)/)[1];
+		}
+		if (id) {
+			ids.push("https://api.osf.io/v2/preprints/" + id + "/?embed=contributors&embed=provider");
+		}
+	}
+	return ids;
+}
+
+function osfAPIImport(text) {
+	// Z.debug(text);
+	let json = JSON.parse(text);
+	let attr = json.data.attributes;
+	let embeds = json.data.embeds;
+	var item = new Zotero.Item("document");
+	// currently we're just doing preprints, but putting this here in case we'll want to handle different OSF
+	// item types in the future
+	// let type = json.data.type
+	item.title = attr.title;
+	item.abstractNote = attr.description;
+	item.date = attr.date_published;
+	item.publisher = embeds.provider.data.attributes.name;
+	if (json.data.links.preprint_doi) {
+		let doi = json.data.links.preprint_doi.replace(/https:\/\/doi\.org\//, "");
+		item.extra = "DOI: " + doi + "\ntype: article";
+	}
+	else {
+		item.extra = "type: article";
+	}
+	item.url = json.data.links.html;
+	for (let tag of attr.tags) {
+		item.tags.push(tag);
+	}
+
+	for (let contributor of embeds.contributors.data) {
+		let author = contributor.embeds.users.data.attributes;
+		if (author.given_name && author.family_name) {
+			// add middle names
+			let givenNames = author.given_name + ' ' + author.middle_names;
+			item.creators.push({ lastName: author.family_name, firstName: givenNames.trim(), creatorType: "author" });
+		}
+		else {
+			item.creators.push({ lastName: author.full_name, creatorType: "author", fieldMode: 1 });
+		}
+	}
+	if (json.data.relationships.primary_file) {
+		let fileID = json.data.relationships.primary_file.links.related.href.replace("https://api.osf.io/v2/files/", "");
+		item.attachments.push({ url: "https://osf.io/download/" + fileID, title: "OSF Preprint", mimeType: "application/pdf" });
+	}
+
+	item.complete();
+}
+
+function scrape(doc, url) {
+	var translator = Zotero.loadTranslator('web');
+	// Embedded Metadata
+	translator.setTranslator('951c027d-74ac-47d4-a107-9c3069ab7b48');
+	// translator.setDocument(doc);
+
+	translator.setHandler('itemDone', function (obj, item) {
+		// TODO adjust if needed:
+		if (item.extra) {
+			item.extra += "\ntype: article";
```

Comment:
@dstillman running into a problem here that looks to me like a bug. Testing e.g. on https://psyarxiv.com/grwd6/ this \only imports the DOI into extra, not `type: article`. If I add a space before type, i.e. `" \ntype: article"` this works (but I think then Zotero/citeproc doesn't pick it up + it should be necessary).
This appears to be a general issue for multi-line entries in Extra -- it fails when importing multiples ([l.114](https://github.com/zotero/translators/pull/2248/files#diff-ba54fd86950c3677ea1d6afed010e23aR114)) as well. 
Tested on Chrome and Firefox. Thoughts?

Reply:
Not a bug. The item type is `document`; CSL `article` maps to Zotero `document` and is therefore removed as redundant. Why do you want to set it?

Reply:
doh! Of course. 

The reason I want to set it is to mark the item as a preprint in the metadata. I'm assuming/hoping that Zotero will get a preprint item type in the future and this should be migrated. `article` is what CSL 1.0.2 explicitly recommend for preprints: https://github.com/citation-style-language/documentation/blob/master/specification.rst#appendix-iii---types 

Thoughts on how to best do this in a forward-looking way?

Reply:
We can add an exception to allow `type: article`, if you think we'll want to migrate such items to the Preprint item type. Presumably not many items are being created with `type: article` for other purposes?

Reply:
Yes, I agree that anything with `type: article` should be moved to Preprint. CSL 1.0.2 has clarified that `type: article` should be used for preprints and working papers. It also adds a generic `type: document`. I would suggest that when Zotero adds a Preprint type (mapped to CSL `article`), it remap Document to CSL `document`. 

For existing Document items _without_ `type: article` in them, I wonder if they should be migrated to Preprint? That would produce consistent citations (e.g., in my library, I store preprints as Document currently because of the type mapping).

Reply:
> For existing Document items without type: article in them, I wonder if they should be migrated to Preprint? That would produce consistent citations (e.g., in my library, I store preprints as Document currently because of the type mapping).

My sense is that most people wouldn't be happy with this. Document is used quite differently.

Reply:
Fair enough.

Reply:
@dstillman I don't know when that changed, but I'm now seeing reports with `type: article` in Extra being auto-converted to `Document` -- that means the above solution for marking preprints doesn't work anymore. Could the previous behavior be re-enabled (or do you have another idea)?

Reply:
@adam3smith: It was [this](https://forums.zotero.org/discussion/comment/369340/#Comment_369340), which you [signed off on](https://forums.zotero.org/discussion/comment/369350/#Comment_369350). The idea here is that an item coming in (from sync or translators) with `Type: <valid CSL type>` should be migrated to the appropriate Zotero type in a future version. This code doesn't bother to exclude the types that already exist now, since those might as well be removed if redundant. But since `article` already exists and we want to use it specifically for a new Preprint item type, we can add an exception.

Reply:
(To be clear, you're only seeing this in the Zotero Beta, right? This was changed since the latest production release.)